pkg/controller/direct/monitoring/roundtrip_test.go

```diff
@@ -0,0 +1,388 @@
+			unimplementedFields.Insert(widgetPath + ".xy_chart.y2_axis")
+			unimplementedFields.Insert(widgetPath + ".xy_chart.timeshift_duration")
+			unimplementedFields.Insert(widgetPath + ".xy_chart.thresholds[].target_axis")
+
+			// This one might be a bug?
+			unimplementedFields.Insert(widgetPath + ".xy_chart.data_sets[].min_alignment_period.nanos")
+
+			unimplementedFields.Insert(widgetPath + ".scorecard.thresholds[].target_axis")
+			unimplementedFields.Insert(widgetPath + ".scorecard.blank_view")
+			unimplementedFields.Insert(widgetPath + ".scorecard.time_series_query.prometheus_query")
+			unimplementedFields.Insert(widgetPath + ".scorecard.time_series_query.output_full_duration")
+
+			unimplementedFields.Insert(widgetPath + ".alert_chart")
+
+			unimplementedFields.Insert(widgetPath + ".time_series_table")
+
+			unimplementedFields.Insert(widgetPath + ".collapsible_group")
+
+			unimplementedFields.Insert(widgetPath + ".pie_chart")
+
+			unimplementedFields.Insert(widgetPath + ".single_view_group")
+
+			unimplementedFields.Insert(widgetPath + ".section_header")
+
+			unimplementedFields.Insert(widgetPath + ".time_series_table")
+
+			unimplementedFields.Insert(widgetPath + ".error_reporting_panel")
+
+			unimplementedFields.Insert(widgetPath + ".incident_list")
+
+			unimplementedFields.Insert(widgetPath + ".text.style")
+
+			unimplementedFields.Insert(widgetPath + ".id")
+		}
+
+		// Remove any output only or known-unimplemented fields
+		clearFields := &ClearFields{
+			Paths: unimplementedFields.Union(outputFields),
+		}
+		visit("", p1.ProtoReflect(), nil, clearFields)
+
+		// Force resource_names to be valid
+		r := &ReplaceFields{}
+		r.Func = func(path string, val protoreflect.Value) (protoreflect.Value, bool) {
+			if strings.HasSuffix(path, ".resource_names[]") {
+				return protoreflect.ValueOfString("projects/" + val.String()), true
+			}
+			return protoreflect.Value{}, false
+		}
+		visit("", p1.ProtoReflect(), nil, r)
+
+		ctx := &MapContext{}
+		k := MonitoringDashboardSpec_FromProto(ctx, p1)
+		if ctx.Err() != nil {
+			t.Fatalf("error mapping from proto to krm: %v", ctx.Err())
+		}
+
+		p2 := MonitoringDashboardSpec_ToProto(ctx, k)
+		if ctx.Err() != nil {
+			t.Fatalf("error mapping from krm to proto: %v", ctx.Err())
+		}
+
+		if diff := cmp.Diff(p1, p2, protocmp.Transform()); diff != "" {
+			t.Logf("p1 = %v", prototext.Format(p1))
+			t.Logf("p2 = %v", prototext.Format(p2))
+			t.Errorf("roundtrip failed; diff:\n%s", diff)
+		}
+	})
+}
+
+func fillWithRandom(t *testing.T, rand *rand.Rand, msg proto.Message) {
+	fillWithRandom0(t, rand, msg.ProtoReflect())
+}
+
+func fillWithRandom0(t *testing.T, rand *rand.Rand, msg protoreflect.Message) {
+	descriptor := msg.Descriptor()
+	if string(descriptor.FullName()) == "google.protobuf.Duration" {
+		count := rand.Intn(10)
+		// Bias to zero
+		if count > 4 {
+			return
+		}
+		seconds := rand.Intn(365 * 24 * 60 * 60)
```

Comment:
do we need to seed `rand` for determinism ?

Reply:
Will rename to randStream.  And we do, the first "real" line of FuzzFromProto